### PR TITLE
feat(buildsystem): add CMake Buildsytem

### PR DESCRIPTION
**Pull request by avinal**
## Description

This pull request adds a new build system based on CMake to this project. This is a big upgrade over decade-old Makefiles. With the new build system, we will be able to build/install/package FOSSology faster and support for many platforms and another build system. 

### Changes

- Added CMake configuration for building, installing, testing, and packaging all the agents
- Refactored source directories to better integrate with the new build system
- Reworked the shell scripts and generated source files to make them more compatible with CMake as well as better in terms of overall compatibility with the latest tools.
- Refactored test source 

## How to test

- Check out the pull request branch 

 ```bash
    git fetch https://github.com/avinal/fossology avinal/feat/buildsystem:buildsystem
    git checkout buildsystem
```

- Make sure you have CMake and Ninja installed

```bash
  sudo apt install cmake ninja-build
```

- Configure CMake, you may add additional variables to modify build
  - `-DOFFLINE=ON` - stops vendor generation
  - `-DAGENTS="agent1;agent2;...;<agent>"` - only configure these agents
  - `-DTESTING=ON` - enables configuring for tests
  - `-DCMAKE_INSTALL_PREFIX=<path>` - changes install prefix for the project
  - `-DMONOPACK=ON` - ununpack and adj2nest are packaged seperately
  - `-GNinja` - uses Ninja Generator instead of  Unix Makefiles
  - `-DCMAKE_BUILD_TYPE=<build-type>` - choose from `Debug`, `Release`, `RelWithDebInfo`, `MinSizeRel`
 
```bash
  mkdir build && cd build
  cmake .. <add-variables-here-if-any>
```

- Run Make or Ninja

```bash
  # For Unix Makefiles, by default, runs without parallel build jobs
  make -j<no-of-parallel-jobs>
  # For Ninja Generator, by default, runs parallel build jobs
  ninja -j<no-of-parallel-jobs>
  # common for all
  cmake --build .
```

- Installing, may need to add `sudo` before the commands

```bash
  # For Unix Makefiles
  make install
  # For Ninja Generator
  ninja install
```

- Testing, on local please consider switching to `fossy` user before running tests

```bash
  # For Unix Makefiles
  make test
  # For Ninja Generator
  ninja test
  # dedicated ctest command *see `ctest --help` for more options*
  ctest --parallel n
```

- Packaging

```bash
  # For Unix Makefiles
  make package
  # For Ninja Generator
  ninja package
```

- Follow the rest of the installation from [here](https://github.com/fossology/fossology/wiki/Install-from-Source#4-run-the-postinstall-script)
- To list all the possible targets 

```bash
  # For Unix Makefiles
  make help
  # For Ninja Generator
  ninja -t targets
```

## Known Bugs/Issue

- While packaging with `make` symbolic links in the packages are accompanied with dereferenced targets too, prefer `ninja` for packaging
- Coverage build may fail for some agents

### Issues to create before merge
- [ ] Fix tests for scheduler, ununpack and delagent units tests.
- [ ] Add requirements for CMake version > 3.13 in Wiki
- [ ] Update install from source instructions in Wiki
- [ ] Add packaging instructions to use `-DCMAKE_INSTALL_PREFIX:PATH="/usr" -DFO_SYSCONFDIR:PATH="/etc/fossology"` before packaging to fix paths.

For More Information please see the blog: https://gsoc.avinal.space and [wiki](https://github.com/avinal/fossology/wiki)

Signed-off-by: Avinal Kumar <avinal.xlvii@gmail.com>

<a href="https://gitpod.io/#https://github.com/fossology/fossology/pull/2075"></a>

### CODE REVIEW

File: src/genvendor/ComposerInstall [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+
```
Comment (GMishx) [inline src/genvendor/ComposerInstall:2]: Please add licensing header.

File: src/genvendor/composer.json [hunk 2]
```diff
@@ -0,0 +1,78 @@
+{
```
Comment (GMishx) [inline src/genvendor/composer.json:1]: Is it required to move `composer.json` from `src/composer.json`?

**Review by GMishx — COMMENTED** [review]
Few suggestions and comments.

File: src/genvendor/composer.json (hunk 2 above)
Comment (avinal) [reply src/genvendor/composer.json:1]: If testing is sorted out then I have alternate plans for it. For now, it must be moved to the build directory. I can refactor the script to use it from the base directory too.

File: install/defconf/fossology.conf.in [hunk 3]
```diff
@@ -41,11 +41,11 @@ localhost[] = * 00 ff
 ; Do not change after installation since PROJECT and PROJECTGROUP is used in installation 
 ; directory names and ownership.
 ; Use PREFIX to install multiple copies of FOSSology.
-PROJECT={$PROJECT}
-PROJECTGROUP={$PROJECTGROUP}
+PROJECT=@FO_PROJECT@
+PROJECTGROUP=@FO_PROJECTGROUP@
```
[the remark this answers is not in the document]
Comment (avinal) [reply install/defconf/fossology.conf.in:45]: Noted

File: src/monk/agent/generator/genSquareVisitor [hunk 4]
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/bash
```
Comment (NicolasToussaint) [inline src/monk/agent/generator/genSquareVisitor:1]: Is there a specific reason for changing to Bash ?
Default shell would be Dash, which is lighter and faster, and I do not see Bash specific code in the file.

File: utils/fo-cleanold [hunk 5]
```diff
@@ -244,8 +241,7 @@ crontab -r -u fossy
 if [ $RMREPO ]; then
    echo "*** Deleting repository ***"
    if [ -d /srv/fossology/repository ]; then
-      rm -rf /srv/fossology
-      if [ $? != 0 ]; then
+      if rm -rf /srv/fossology; then
```
Comment (NicolasToussaint) [inline utils/fo-cleanold:244]: There is a logic inversion here, we should have: ` if ! rm -rf /srv/fossology; then`

File: utils/fo-cleanold [hunk 6]
```diff
@@ -201,31 +201,28 @@ fi
...
+      rm -rf "$dir"
    fi
 done
 
 ## Database
 if [ $RMDB ]; then
    # first check that the db is running
-   su postgres -c 'echo \\q|psql'
-   if [ $? != 0 ]; then
+   if su postgres -c 'echo \\q|psql'; then
```
Comment (NicolasToussaint) [inline utils/fo-cleanold:217]: Logic inversion, we should have: ` if ! [...]; then`

File: utils/fo-cleanold [hunk 7]
```diff
@@ -201,31 +201,28 @@ fi
...
       # and next that the db exists
-      su postgres -c 'psql -l' |grep -q fossology
-      if [ $? = 0 ]; then
+      if ! su postgres -c 'psql -l' | grep -q fossology; then
          echo "*** Deleting database ***"
          # stop all users of the fossology db
          pkill -f -u postgres fossy || true
-         su postgres -c 'echo "drop database fossology"|psql'
-         if [ $? != 0 ]; then
+         if su postgres -c 'echo "drop database fossology"|psql'; then
```
Comment (NicolasToussaint) [inline utils/fo-cleanold:225]: Logic inversion, we should have: ` if ! [...]; then`

File: utils/fo-cleanold [hunk 8]
```diff
@@ -258,16 +254,14 @@ fi
 if [ $RMUSER ]; then
    echo "*** Deleting user and group ***"
    if grep -q "^fossy:" /etc/passwd; then
-      userdel -r fossy
-      if [ $? != 0 ]; then
+      if userdel -r fossy; then
```
Comment (NicolasToussaint) [inline utils/fo-cleanold:257]: Logic inversion, we should have: ` if ! [...]; then`

File: utils/fo-cleanold [hunk 9]
```diff
@@ -258,16 +254,14 @@ fi
...
          echo "ERROR: failed to delete user"
          exit 1
       fi
    fi
 
    if grep -q "^fossy:" /etc/group; then
-      groupdel fossy
-      if [ $? != 0 ]; then
+      if groupdel fossy; then
```
Comment (NicolasToussaint) [inline utils/fo-cleanold:264]: Logic inversion, we should have: ` if ! [...]; then`

File: src/monk/agent/generator/genSquareVisitor (hunk 4 above)
Comment (avinal) [reply src/monk/agent/generator/genSquareVisitor:1]: There was some permission issue earlier and I thought this was causing them. I have fixed them now. I will revert it back to `sh`.

**Comment by avinal** [comment]
Hello @NicolasToussaint, I did the changes in the `fo-cleanold` according to the latest `shellcheck` guidelines but it seems that some commands are broken after the changes. So I will be reverting them back to old. Although, thanks for pointing the logic errors :)

**Comment by avinal** [comment]
Hi @GMishx, I was trying to implement the copyright add functionality to the cpack. But there is a catch, the variables that the script suggested [here](https://gitlab.kitware.com/cmake/cmake/-/issues/21832) is using features not present in the CMake versions we are targeting, (3.10 opposed to 3.19). The one way we can make it work is by running CPack exclusively with CMake 3.19+. What are your thoughts on this?

**Comment by GMishx** [comment]
@avinal we need some help. On Ubuntu Bionic Beaver (18.04), the CMake version is 3.10.2
While configuring fossology, we get following error:
```
CMake Error at src/copyright/agent/CMakeLists.txt:36 (target_link_libraries):
  Object library target "copyright.cc.o" may not link to anything.
```

And if I comment the line 36-38, this error disappears and an error on `src/wget_agent/agent/CMakeLists.txt:47` and `src/wget_agent/agent/CMakeLists.txt:53` pop up saying `checksum` is an object library and can not be exported.

On CMake version 3.18.4 and 3.23, we do not face any such issue.
Can you please help us solve the mentioned 2 issues?

**Comment by avinal** [comment]
Hello Gaurav, I will see it tonight and let you know. Most probably some CMake configurations has been changed. Wget depends on checksum object from a different agent.

**Comment by GMishx** [comment]
> Hello Gaurav, I will see it tonight and let you know. Most probably some CMake configurations has been changed. Wget depends on checksum object from a different agent.

Thanks!
And yes, checksum comes from `src/ununpack/agent/CMakeLists.txt:100`

**Comment by GMishx** [comment]
Once everything is reviewed, we can remove the `Makefile`s from the source.

**Comment by avinal** [comment]
> Once everything is reviewed, we can remove the `Makefile`s from the source.

Great 🥳 , Thanks

**Comment by avinal** [comment]
Finally 😍

**Pull request merged**